vm/src/object/core.rs

```diff
@@ -849,7 +861,23 @@ impl<'a, T: PyObjectPayload> From<&'a Py<T>> for &'a PyObject {
 impl Drop for PyObjectRef {
     #[inline]
     fn drop(&mut self) {
```

Comment:
Because `PyObjectRef::drop` is very small, I'd like to suggest to add a whole debug drop , rather than adding a lot bigger debug features inside of the functions.

```rust
#[cfg(debug_assertions)]
fn drop(&mut self) {
    // debug drop here
}

#[cfg(not(debug_assertions))
#[inline]
fn drop(&mut self) {
    // production drop here
}
```

Reply:
> Because `PyObjectRef::drop` is very small, I'd like to suggest to add a whole debug drop , rather than adding a lot bigger debug features inside of the functions.
> 
> ```rust
> #[cfg(debug_assertions)]
> fn drop(&mut self) {
>     // debug drop here
> }
> 
> #[cfg(not(debug_assertions))
> #[inline]
> fn drop(&mut self) {
>     // production drop here
> }
> ```

Doing the refactor right now, after I done refactor, should I amend&force push or create a new commit?

Reply:
> Because `PyObjectRef::drop` is very small, I'd like to suggest to add a whole debug drop , rather than adding a lot bigger debug features inside of the functions.
> 
> ```rust
> #[cfg(debug_assertions)]
> fn drop(&mut self) {
>     // debug drop here
> }
> 
> #[cfg(not(debug_assertions))
> #[inline]
> fn drop(&mut self) {
>     // production drop here
> }
> ```

Also I refactor GC PR on top of this drop check, which, you can see, make drop() a much larger function, should it be spilted the same way?
```rust
impl Drop for PyObjectRef {
    #[inline]
    fn drop(&mut self) {
        #[cfg(debug_assertions)]
        if *self.0.is_drop.lock() {
            error!(
                "Double drop on PyObjectRef with typeid={:?}(Type={:?})",
                self.0.typeid,
                ID2TYPE
                    .lock()
                    .expect("can't read ID2TYPE")
                    .get(&self.0.typeid),
            );
            return;
        }
        #[cfg(feature = "gc")]
        {
            let stat = self.decrement();
            match stat {
                GcStatus::ShouldDrop => unsafe {
                    *self.0.is_drop.lock() = true;
                    PyObject::drop_slow(self.ptr);
                },
                GcStatus::ShouldDropOnly | GcStatus::BufferedDrop => unsafe {
                    *self.0.is_drop.lock() = true;
                    PyObject::drop_only(self.ptr);
                },
                GcStatus::ShouldKeep | GcStatus::DoNothing => (),
            }
        }
        #[cfg(not(feature = "gc"))]
        if self.0.ref_count.dec() {
            #[cfg(debug_assertions)]
            {
                *self.0.is_drop.lock() = true;
            }
            unsafe { PyObject::drop_slow(self.ptr) }
        }
    }
}
```